docs/BOOT_LOADER_SPECIFICATION.md

```diff
@@ -23,72 +23,106 @@ includes a boot loader.
   firmware itself
 * OS installer developers, to create appropriate partitions and set up the
-  initial boot loader configuration
-* Distribution developers, to create appropriate configuration snippets when
+  initial boot loader menu entries
+* Distribution developers, to create appropriate menu entry snippets when
   installing or updating kernel packages
 * UI developers, to implement user interfaces that list and select among the
   available boot options
 
-## The Boot Partition
+## The Partitions
 
 Everything described below is located on one or two partitions. The boot loader
-or user-space programs reading the boot loader configuration should locate them
+or user-space programs reading the boot loader menu entries should locate them
 in the following manner:
 
 * On disks with an MBR partition table:
 
-  * The boot partition — partition with the type ID of 0xEA — shall be used
-    for boot loader configuration and entries.
+  * The boot partition — a partition with the type ID of `0xEA` — shall be used
+    as single location for boot loader menu entries.
 
 * On disks with GPT (GUID Partition Table)
 
-  * The EFI System Partition (ESP for short) — a partition with GPT type GUID
-    of `c12a7328-f81f-11d2-ba4b-00a0c93ec93b` — should be used for boot loader
-    configuration and boot entries.
+  * The EFI System Partition (ESP for short) — a partition with a GPT type GUID
+    of `c12a7328-f81f-11d2-ba4b-00a0c93ec93b` — may be used as one of two locations for
+    boot loader menu entries.
 
   * Optionally, an Extended Boot Loader Partition (XBOOTLDR partition for
     short) — a partition with GPT type GUID of
-    `bc13c2ff-59e6-4262-a352-b275fd6f7172` — may be used as an additional
-    location for boot loader entries. This partition must be located on the
-    same disk as the ESP.
+    `bc13c2ff-59e6-4262-a352-b275fd6f7172` — may be used as the second of two
+    locations for boot loader menu entries. This partition must be located on
+    the same disk as the ESP.
 
-In the text below, `$BOOT` will be used to refer to (the root of) the first of
-the two partitions (the boot partition on MBR disks and the ESP on GPT disks),
-and `$XBOOTLDR` will be used to refer to (the root of) the optional second
-partition.
+There may be at most one partition of each of the types listed above on the
+same disk.
 
-An installer for the operating system should use this logic when selecting or
+**Note:** _These partitions are **shared** among all OS installations on the
+same disk. Instead of maintaining one boot partition per installed OS (as
+`/boot/` was traditionally handled), all installed OSes use the same place for
+boot loader menu entries._
+
+For systems where the firmware is able to read file systems directly, the ESP
+must — and the MBR boot and GPT XBOOTLDR partition should — be a file system
+readable by the firmware. For most systems this means VFAT (16 or 32
```

Comment:
Would [efifs](https://github.com/pbatard/efifs) drivers satisfy this requirement? The pre-boot environment would thus be able to read most any file system, but of course it's limited since these fs drivers need to be signed for Secure Boot. That means we'd potentially get this explosion of distro signed drivers, having to encourage everyone to follow such a paradigm just to be certain they could read whatever volume format $BOOT happens to be. Permissive or strict?

Reply:
I think using any other FS than FAT for XBOOTLDR is stupid. General purpose distros shouldn't do that. People who don't have to care about the general purpose and only focus on specific devices (i.e. devices without need for multi-boot) otoh can choose something else if they really like to.

Reply:
A significant advantage of using Btrfs (or ZFS) for this purpose is pooled space. You don't have to worry about the problem of $BOOT size, it's all shared space with $ROOT. I wonder if the spec should permit multiple XBOOTLDR, in effect pooling them? This means we can start with 1G $BOOT and if later on it's not big enough, just add another one instead of having to reprovision.

e.g. consider the case of https://github.com/systemd/systemd/pull/23841 pinning kernels+initramfs to $BOOT so that older snapshots can be booted. Estimating the size of $BOOT isn't exactly straightforward in advance, so in these cases we tend to over provision to avoid premature reprovision.

Reply:
@cmurf well, it am not convinced it's that hard to come up with a good estimation how much space you'll need and then double that or so. But sure, if this really really becomes an issue we can revisit this one day, but I am not convinced we are at that day yet. 

I mean, yes, when ESPs first appeared on the scene people didn't know how we'd like to use them today, so we added XBOOTLDR to address that. But we know pretty well what we want to use XBOOTLDR for for now, which makes it much easier to estimate a size that might work.

One day, when we realize that maybe we want to use it differently we can revisit things, but doing so ahead of time without knowing the problems the future brings is premature.